src/WinRT.Runtime/ExceptionHelpers.cs

```diff
@@ -151,68 +151,75 @@ private static Exception GetExceptionForHR(int hr, bool useGlobalErrorState, out
-                        // For cross language WinRT exceptions, general information will be available in the description,
-                        // which is populated from IRestrictedErrorInfo::GetErrorDetails and more specific information will be available
-                        // in the restrictedError which also comes from IRestrictedErrorInfo::GetErrorDetails. If both are available, we
-
-                        // need to concatinate them to produce the final exception message.
-                        if (!string.IsNullOrEmpty(description))
+                        restrictedErrorInfoToSave = ObjectReference<IUnknownVftbl>.FromAbi(restrictedErrorInfoValuePtr, IID.IID_IRestrictedErrorInfo);
+                        ABI.WinRT.Interop.IRestrictedErrorInfoMethods.GetErrorDetails(restrictedErrorInfoValuePtr, out description, out int hrLocal, out restrictedError, out restrictedCapabilitySid);
+                        restrictedErrorReference = ABI.WinRT.Interop.IRestrictedErrorInfoMethods.GetReference(restrictedErrorInfoValuePtr);
+                        if (hr == hrLocal)
                         {
-                            errorMessage += description;
+                            // For cross language WinRT exceptions, general information will be available in the description,
+                            // which is populated from IRestrictedErrorInfo::GetErrorDetails and more specific information will be available
+                            // in the restrictedError which also comes from IRestrictedErrorInfo::GetErrorDetails. If both are available, we
 
-                            if (!string.IsNullOrEmpty(restrictedError))
+                            // need to concatinate them to produce the final exception message.
+                            if (!string.IsNullOrEmpty(description))
                             {
-                                errorMessage += Environment.NewLine;
+                                errorMessage += description;
+
+                                if (!string.IsNullOrEmpty(restrictedError))
+                                {
+                                    errorMessage += Environment.NewLine;
+                                }
                             }
-                        }
 
-                        if (!string.IsNullOrEmpty(restrictedError))
-                        {
-                            errorMessage += restrictedError;
+                            if (!string.IsNullOrEmpty(restrictedError))
+                            {
+                                errorMessage += restrictedError;
+                            }
                         }
                     }
                 }
+                catch (Exception e)
+                {
+                    // If we fail to get the error info or the exception from it,
+                    // we fallback to using the hresult to create the exception.
+                    Debug.Assert(false, e.Message, e.StackTrace);

```

Comment:
Is this catch (and the one in the other helper below) expected to only come from a specific native call? I'm wondering if, assuming that's the case, we can just check the HRESULT instead and go to the fallback path, rather than having it throw and catch it, which would be much less efficient. Kinda like the optimization we did in `CreateCcwForMarshalingForABI` (or whatever that was called). Thoughts?

Reply:
This is a catch all for we shouldn't accidentally throw a diff exception from what we are actually trying to throw.  i.e. previously when we were trying to get ThisPtr, we ran into ObjectDisposedExcepton and that replaced the exception we were throwing.  Instead this should be best-effort and if we fail, we still throw the original exception based on hr.

Reply:
That makes sense. I'm just wondering what cases could possibly still be there? Like, if we did end up in such a scenario, would we not watch to detect that as soon as possible, as that would indicate a bug in CsWinRT? Or, could that not indicate some potentially bigger issue (like, something breaking somewhere) that we wouldn't want to hide?

Reply:
yea I was debating that which is why I put in a Debug.Assert.  At the same time, we still probably shouldn't replace the original exception as that just makes it harder to diagnose the app issue.  The worst case here is app may get less exception context and I think we would get feedback on that as we have in the past if something goes wrong.

Reply:
I guess my concerns are:
- How can we be sure if we got some other exception here that we just ignore, the process is in a valid state?
- How would we ever detect such scenarios if all release builds of CsWinRT will just ignore the exception?

Reply:
Given this function is specific to error propagation rather than a core functionality, we are saying that even if it is due to the process being in a bad state, we shouldn't hide the original issue which could be what leads to that.  This is technically similar to how the OS error propagation APIs work as they are best effort for diagnostics.

We are saying that if there is an actual issue in error propagation implementation of CsWinRT, we would need to catch it with our tests which is now more thorough, and we would also learn from crash dumps losing diagnostic context.

Reply:
Also the other thing is when we throw other unrelated exceptions during exception propagation, it can become misleading and confusing what is happening.

Reply:
That seems fair. Idea, could we perhaps add the inner exception to a well-known item in `Data` in these cases? So that it's not lost entirely, and it'd be able to get info out of it if needed? Rather than just completely swallowing it.

Reply:
I think, we might be able to do that.

Reply:
Added it for this scenario.  The others we probably can't do much as that happens when converting to a native error.  But those are less complicated too and less likely to be an issue that we don't catch in our tests.